src/android/l2cap_channel.rs

```diff
@@ -142,79 +152,58 @@ impl Drop for L2capCloser {
 }
 
 pub struct L2capChannelReader {
-    stream: Receiver<Vec<u8>>,
+    stream: piper::Reader,
     closer: Arc<L2capCloser>,
 }
 
 impl L2capChannelReader {
-    #[inline]
-    pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
-        let packet = self
-            .stream
-            .recv()
-            .await
-            .map_err(|_| Error::new(ErrorKind::ConnectionFailed, None, "channel is closed"))?;
-
-        if packet.len() > buf.len() {
-            return Err(Error::new(ErrorKind::InvalidParameter, None, "Buffer is too small"));
-        }
-
-        buf[..packet.len()].copy_from_slice(&packet);
-
-        Ok(packet.len())
-    }
-
-    #[inline]
-    pub fn try_read(&mut self, buf: &mut [u8]) -> Result<usize> {
-        let packet = self.stream.try_recv().map_err(|e| match e {
-            TryRecvError::Empty => Error::new(ErrorKind::NotReady, None, "no received packet in queue"),
-            TryRecvError::Closed => Error::new(ErrorKind::ConnectionFailed, None, "channel is closed"),
-        })?;
-
-        if packet.len() > buf.len() {
-            return Err(Error::new(ErrorKind::InvalidParameter, None, "Buffer is too small"));
-        }
-
-        buf[..packet.len()].copy_from_slice(&packet);
-
-        Ok(packet.len())
-    }
-
     pub async fn close(&mut self) -> Result<()> {
         self.closer.close();
         Ok(())
     }
 }
 
+impl AsyncRead for L2capChannelReader {
+    fn poll_read(mut self: pin::Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<std::io::Result<usize>> {
+        let stream = pin::pin!(&mut self.stream);
+        stream.poll_read(cx, buf)
+    }
+}
+
 impl fmt::Debug for L2capChannelReader {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.write_str("L2capChannelReader")
     }
 }
 
 pub struct L2capChannelWriter {
-    stream: Sender<Vec<u8>>,
+    stream: piper::Writer,
     closer: Arc<L2capCloser>,
 }
 
 impl L2capChannelWriter {
-    pub async fn write(&mut self, packet: &[u8]) -> Result<()> {
-        self.stream
-            .send(packet.to_vec())
-            .await
-            .map_err(|_| Error::new(ErrorKind::ConnectionFailed, None, "channel is closed"))
+    pub async fn close(&mut self) -> Result<()> {
+        self.closer.close();
+        Ok(())
     }
+}
 
-    pub fn try_write(&mut self, packet: &[u8]) -> Result<()> {
-        self.stream.try_send(packet.to_vec()).map_err(|e| match e {
-            TrySendError::Closed(_) => Error::new(ErrorKind::ConnectionFailed, None, "channel is closed"),
-            TrySendError::Full(_) => Error::new(ErrorKind::NotReady, None, "No buffer space for write"),
-        })
+impl AsyncWrite for L2capChannelWriter {
+    fn poll_write(mut self: pin::Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<std::io::Result<usize>> {
+        let stream = pin::pin!(&mut self.stream);
+        let ret = stream.poll_write(cx, buf);
+        ret
     }
```

Comment:
Could we still expose this non-blocking try_write or is it available via some traits?

Reply:
I've been using it as writer.[write_all](https://github.com/smol-rs/futures-lite/blob/3444c7a2f8fe423c49d08bdedfa5b6b188fa57b5/src/io.rs#L2834)(&buf).await, from future-lite's `pub trait AsyncWriteExt: AsyncWrite`.
I think it does repeated poll_write(&whatevers_left_buf) until it returns Ok((0)) (written bytes) or Err.
Worked no prob for writing 100kb bytes buf

Reply:
Ah, thanks. now_or_never() was what I was looking for.